File: surge_detector.py
from collections import deque
import numpy as np
from datetime import datetime, timedelta
import threading

class SurgeDetector:
    def __init__(self, window_size=60, threshold_std=2.0):
        """
        Initialize the surge detector.
        
        Args:
            window_size (int): Size of the rolling window in minutes
            threshold_std (float): Number of standard deviations for spike detection
        """
        self.window_size = window_size
        self.threshold_std = threshold_std
        self.values = deque(maxlen=window_size)
        self.timestamps = deque(maxlen=window_size)
        self.spikes = deque(maxlen=100)  # Keep track of last 100 spikes
        self.lock = threading.Lock()

    def add_datapoint(self, value):
        """
        Add a new datapoint and check for spike.
        
        Args:
            value (float): The value to add
            
        Returns:
            bool: True if spike detected, False otherwise
        """
        with self.lock:
            current_time = datetime.now()
            self.values.append(value)
            self.timestamps.append(current_time)
            
            # Need at least 5 values for meaningful statistics
            if len(self.values) < 5:
                return False
            
            # Calculate statistics
            mean = np.mean(self.values)
            std = np.std(self.values)
            threshold = mean + (self.threshold_std * std)
            
            # Check for spike
            is_spike = value > threshold
            if is_spike:
                self.spikes.append((current_time, value))
            
            return is_spike

    def get_current_average(self):
        """Get the current rolling average."""
        with self.lock:
            if not self.values:
                return 0
            return float(np.mean(self.values))

    def get_recent_values(self):
        """Get recent values and timestamps for plotting."""
        with self.lock:
            return [
                {
                    'timestamp': ts.isoformat(),
                    'value': val
                }
                for ts, val in zip(self.timestamps, self.values)
            ]

    def get_recent_spikes(self):
        """Get recent spike events."""
        with self.lock:
            return [
                {
                    'timestamp': ts.isoformat(),
                    'value': val
                }
                for ts, val in self.spikes
            ]

    def get_threshold(self):
        """Get current threshold value."""
        with self.lock:
            if len(self.values) < 5:
                return None
            mean = np.mean(self.values)
            std = np.std(self.values)
            return float(mean + (self.threshold_std * std))

    def set_threshold(self, threshold_std):
        """Update the threshold multiplier."""
        if threshold_std <= 0:
            raise ValueError("Threshold must be positive")
        with self.lock:
            self.threshold_std = threshold_std

File: surgeDetector.js
const moment = require('moment');

class SurgeDetector {
    constructor(windowSize = 60, thresholdStd = 2.0) {
        this.windowSize = windowSize;
        this.thresholdStd = thresholdStd;
        this.handleData = new Map(); // Store data per handle
        this.maxSpikes = 100;
    }

    initializeHandleData(handle) {
        if (!this.handleData.has(handle)) {
            this.handleData.set(handle, {
                values: [],
                timestamps: [],
                spikes: []
            });
        }
        return this.handleData.get(handle);
    }

    addDatapoint(handle, value) {
        const currentTime = moment();
        const data = this.initializeHandleData(handle);

        // Add new value and timestamp
        data.values.push(value);
        data.timestamps.push(currentTime);

        // Remove old values outside the window
        const cutoffTime = moment().subtract(this.windowSize, 'minutes');
        while (data.timestamps.length > 0 && data.timestamps[0].isBefore(cutoffTime)) {
            data.timestamps.shift();
            data.values.shift();
        }

        // Need at least 5 values for meaningful statistics
        if (data.values.length < 5) {
            return false;
        }

        // Calculate statistics
        const mean = this.calculateMean(data.values);
        const std = this.calculateStd(data.values, mean);
        const threshold = mean + (this.thresholdStd * std);

        // Check for spike
        const isSpike = value > threshold;
        if (isSpike) {
            data.spikes.push({ timestamp: currentTime, value });
            // Keep only the latest spikes
            if (data.spikes.length > this.maxSpikes) {
                data.spikes.shift();
            }
        }

        return isSpike;
    }

    calculateMean(values) {
        return values.reduce((sum, val) => sum + val, 0) / values.length;
    }

    calculateStd(values, mean) {
        const squareDiffs = values.map(value => {
            const diff = value - mean;
            return diff * diff;
        });
        const avgSquareDiff = this.calculateMean(squareDiffs);
        return Math.sqrt(avgSquareDiff);
    }

    getCurrentAverage(handle) {
        const data = this.handleData.get(handle);
        return data && data.values.length ? this.calculateMean(data.values) : 0;
    }

    getRecentValues(handle) {
        const data = this.handleData.get(handle);
        if (!data) return [];
        return data.values.map((value, index) => ({
            timestamp: data.timestamps[index].toISOString(),
            value
        }));
    }

    getRecentSpikes(handle) {
        const data = this.handleData.get(handle);
        if (!data) return [];
        return data.spikes.map(spike => ({
            timestamp: spike.timestamp.toISOString(),
            value: spike.value
        }));
    }

    getThreshold(handle) {
        const data = this.handleData.get(handle);
        if (!data || data.values.length < 5) {
            return null;
        }
        const mean = this.calculateMean(data.values);
        const std = this.calculateStd(data.values, mean);
        return mean + (this.thresholdStd * std);
    }

    setThreshold(thresholdStd) {
        if (thresholdStd <= 0) {
            throw new Error("Threshold must be positive");
        }
        this.thresholdStd = thresholdStd;
    }
}

module.exports = SurgeDetector;